VariantShredding.md

```diff
@@ -25,276 +25,302 @@
+| INVALID                                                                            | `{"event_type": "login"}`         | non-null      | null                           | `login`                              | null                         | 1729795057774                      | INVALID: Shredded field is present in `value` |
+
+## Nesting
 
+The `typed_value` associated with any Variant `value` field can be any shredded type according to the rules above.
 
-The corresponding Parquet schema with “a” and “b” as leaf types is:
+For example, the `event` object above may also shred sub-fields as object (`location`) or array (`tags`).
 
 ```
-optional group variant_col {
- required binary metadata;
- optional binary variant_value;
- optional group array (LIST) {
-  repeated group list {
-   optional group element {
-    optional binary variant_value;
-    optional group object {
-     optional group a {
-      optional binary variant_value;
-      optional group array (LIST) {
-       repeated group list {
-        optional group element {
-         optional int64 typed_value;
-         optional binary variant_value;
+optional group event (VARIANT) {
+  required binary metadata;
+  optional binary value;
+  optional group typed_value {
+    required group event_type {
+      optional binary value;
+      optional binary typed_value (STRING);
+    }
+    required group event_ts {
+      optional binary value;
+      optional int64 typed_value (TIMESTAMP(true, MICROS));
+    }
+    required group location {
+      optional binary value;
+      optional group typed_value {
+        required group latitude {
+          optional binary value;
+          optional double typed_value;
+        }
+        required group longitude {
+          optional binary value;
+          optional double typed_value;
+        }
+      }
+    }
+    required group tags {
+      optional binary value;
+      optional group typed_value (LIST) {
+        repeated group list {
+          required group element {
+            optional binary value;
+            optional binary typed_value (STRING);
+          }
         }
-       }
       }
-     }
-     optional group b {
-      optional int64 typed_value;
-      optional binary variant_value;
-     }
     }
-   }
   }
- }
 }
 ```
 
-In the above example schema, if “a” is an array containing a mix of integer and non-integer values, the engine will shred individual elements appropriately into either `typed_value` or `variant_value`.
-If the top-level Variant is not an array (for example, an object), the engine cannot shred the value and it will store it in the top-level `variant_value`.
-Similarly, if "a" is not an array, it will be stored in the `variant_value` under "a".
+# Data Skipping
 
-Consider the following example:
+Statistics for `typed_value` columns can be used for file, row group, or page skipping when `value` is always null (missing).
```

Comment:
Do we need to specify "null" vs "variant null" I get a little confused sometimes in the doc.

Reply:
Isn't "null (missing)" clear that it is not variant null? Missing is only used to mean one form of null in the text.